axelor-account/src/main/java/com/axelor/apps/account/service/moveline/MoveLineTaxServiceImpl.java

```diff
@@ -299,30 +299,69 @@ public int getVatSystem(Move move, MoveLine moveline) throws AxelorException {
       return;
     }
-    for (MoveLine moveline : move.getMoveLineList()) {
-      if (moveline.getAccount() != null
-          && moveline.getAccount().getAccountType() != null
-          && AccountTypeRepository.TYPE_TAX.equals(
-              moveline.getAccount().getAccountType().getTechnicalTypeSelect())
-          && !move.getMoveLineList().stream()
-              .filter(
-                  ml ->
-                      moveLineToolService.isEqualTaxMoveLine(
-                          moveline.getAccount(),
-                          moveline.getTaxLine(),
-                          moveline.getVatSystemSelect(),
-                          moveline.getId(),
-                          ml))
-              .collect(Collectors.<MoveLine>toList())
-              .isEmpty()) {
+    for (MoveLine moveLine : move.getMoveLineList()) {
+      if (this.isMoveLineTaxAccount(moveLine) && this.isDuplicateTaxMoveLine(move, moveLine)) {
         throw new AxelorException(
             TraceBackRepository.CATEGORY_NO_VALUE,
             I18n.get(AccountExceptionMessage.SAME_TAX_MOVE_LINES));
       }
     }
   }
+
+  @Override
+  public boolean isMoveLineTaxAccount(MoveLine moveLine) {
+    return moveLine.getAccount() != null
+        && moveLine.getAccount().getAccountType() != null
+        && AccountTypeRepository.TYPE_TAX.equals(
+            moveLine.getAccount().getAccountType().getTechnicalTypeSelect());
+  }
+
+  protected boolean isDuplicateTaxMoveLine(Move move, MoveLine moveLine) {
+    return move.getMoveLineList().stream()
+        .anyMatch(
+            ml ->
+                moveLineToolService.isEqualTaxMoveLine(
+                    moveLine.getAccount(),
+                    moveLine.getTaxLine(),
+                    moveLine.getVatSystemSelect(),
+                    moveLine.getId(),
+                    ml));
+  }
+
+  @Override
+  public void checkEmptyTaxLines(List<MoveLine> moveLineList) throws AxelorException {
+    List<Long> moveLineWithoutTaxList = this.getMoveLinesWithoutTax(moveLineList);
+
+    if (ObjectUtils.notEmpty(moveLineWithoutTaxList)) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_INCONSISTENCY,
+          String.format(
+              I18n.get(AccountExceptionMessage.MOVE_LINE_TAX_LINE_MISSING),
+              moveLineWithoutTaxList));
+    }
+  }
+
+  protected List<Long> getMoveLinesWithoutTax(List<MoveLine> moveLineList) {
+    List<Long> moveLineWithoutTaxList = new ArrayList<>();
+
+    if (CollectionUtils.isEmpty(moveLineList)) {
+      return null;
+    }
+
+    for (MoveLine moveLine : moveLineList) {
+      if (moveLine.getMove() != null
+          && this.isMoveLineTaxAccount(moveLine)
+          && Lists.newArrayList(
+                  MoveRepository.FUNCTIONAL_ORIGIN_PURCHASE, MoveRepository.FUNCTIONAL_ORIGIN_SALE)
+              .contains(moveLine.getMove().getFunctionalOriginSelect())
+          && moveLine.getTaxLine() == null) {
```

Comment:
I think that part of the condition can be factorized with the one used in `MoveLineAttrsServiceImpl.addTaxLineRequired`